File: src/Kela.Api/Media/MediaProcessor.cs
using System.Diagnostics;

namespace Kela.Api.Media;

/// <summary>
/// Video dosyalarını diske kaydeder, ffprobe ile süreyi okur ve
/// ffmpeg ile videodan otomatik thumbnail (1. saniyedeki kare) çıkarır.
/// </summary>
public sealed class MediaProcessor
{
    private readonly string _uploadsRoot;

    public MediaProcessor(IWebHostEnvironment env, IConfiguration config)
    {
        _uploadsRoot = Path.GetFullPath(
            config["Uploads:Root"] ?? Path.Combine(env.ContentRootPath, "uploads"));
    }

    public string UploadsRoot => _uploadsRoot;
    public string VideosDir => Path.Combine(_uploadsRoot, "videos");
    public string ThumbnailsDir => Path.Combine(_uploadsRoot, "thumbnails");

    /// <summary>Videoyu diske yazar, süresini ve thumbnail'ini üretir.</summary>
    /// <returns>(videoUrlPath, thumbnailUrlPath, durationSeconds)</returns>
    public (string VideoPath, string ThumbnailPath, int DurationSeconds) SaveVideo(Stream stream, string originalName)
    {
        Directory.CreateDirectory(VideosDir);
        Directory.CreateDirectory(ThumbnailsDir);

        var ext = Path.GetExtension(originalName).ToLowerInvariant();
        if (string.IsNullOrWhiteSpace(ext) || !IsVideoExtension(ext))
        {
            ext = ".mp4";
        }

        var id = Guid.NewGuid().ToString("N");
        var videoName = id + ext;
        var thumbName = id + ".jpg";

        var videoFile = Path.Combine(VideosDir, videoName);
        var thumbFile = Path.Combine(ThumbnailsDir, thumbName);

        using (var file = new FileStream(videoFile, FileMode.Create, FileAccess.Write, FileShare.None, 1024 * 1024))
        {
            stream.CopyTo(file);
        }

        var duration = ProbeDuration(videoFile);
        ExtractThumbnail(videoFile, thumbFile);

        return ($"/uploads/videos/{videoName}", $"/uploads/thumbnails/{thumbName}", duration);
    }

    /// <summary>DB'deki URL yolu (<c>/uploads/...</c>) → fiziksel dosya yolu.</summary>
    public string? ResolvePhysicalPath(string urlPath)
    {
        if (string.IsNullOrWhiteSpace(urlPath))
        {
            return null;
        }

        // DB'deki yol "/uploads/videos/x.mp4" biçimindedir; _uploadsRoot zaten "uploads" köküdür,
        // bu nedenle öndeki "/uploads/" parçasını ayıkla (yoksa çift önek oluşur).
        var relative = urlPath.TrimStart('/').Replace('/', Path.DirectorySeparatorChar);
        if (relative.StartsWith($"uploads{Path.DirectorySeparatorChar}", StringComparison.OrdinalIgnoreCase))
        {
            relative = relative[($"uploads{Path.DirectorySeparatorChar}").Length..];
        }

        var full = Path.GetFullPath(Path.Combine(_uploadsRoot, relative));
        var root = Path.GetFullPath(_uploadsRoot) + Path.DirectorySeparatorChar;
        return full.StartsWith(root, StringComparison.OrdinalIgnoreCase) && File.Exists(full) ? full : null;
    }

    public static string ContentTypeFor(string urlPath)
    {
        var ext = Path.GetExtension(urlPath).ToLowerInvariant();
        return ext switch
        {
            ".mp4" => "video/mp4",
            ".webm" => "video/webm",
            ".ogg" => "video/ogg",
            ".mov" => "video/quicktime",
            ".m4v" => "video/x-m4v",
            ".mkv" => "video/x-matroska",
            ".avi" => "video/x-msvideo",
            _ => "application/octet-stream",
        };
    }

    private static bool IsVideoExtension(string ext) => ext is
        ".mp4" or ".webm" or ".ogg" or ".mov" or ".m4v" or ".mkv" or ".avi" or ".mpg" or ".mpeg";

    /// <summary>ffprobe ile süre (saniye, aşağı yuvarlanır).</summary>
    private static int ProbeDuration(string videoFile)
    {
        var psi = new ProcessStartInfo("ffprobe")
        {
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true,
            ArgumentList = { "-v", "error", "-show_entries", "format=duration", "-of", "csv=p=0", videoFile },
        };

        using var proc = Process.Start(psi);
        if (proc is null)
        {
            return 0;
        }

        var output = proc.StandardOutput.ReadToEnd().Trim();
        proc.WaitForExit(30_000);

        return double.TryParse(output, System.Globalization.CultureInfo.InvariantCulture, out var seconds)
            ? (int)Math.Round(seconds)
            : 0;
    }

    /// <summary>ffmpeg ile videonun 1. saniyesindeki kareyi thumbnail olarak kaydeder.</summary>
    private static void ExtractThumbnail(string videoFile, string thumbFile)
    {
        var psi = new ProcessStartInfo("ffmpeg")
        {
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true,
            ArgumentList =
            {
                "-i", videoFile,
                "-ss", "00:00:01",
                "-vframes", "1",
                "-vf", "scale=640:-2",
                "-y", thumbFile,
            },
        };

        using var proc = Process.Start(psi);
        if (proc is null)
        {
            return;
        }

        // ffmpeg progress stderr'e yazar; ReadToEnd okuyana kadar çıkmayabilir.
        _ = proc.StandardError.ReadToEndAsync();
        if (!proc.WaitForExit(60_000))
        {
            try { proc.Kill(entireProcessTree: true); } catch { /* boş ver */ }
        }
    }
}

File: laravel/app/Infrastructure/Media/MediaProcessor.php
<?php

namespace App\Infrastructure\Media;

use Illuminate\Support\Facades\Storage;
use Illuminate\Support\Str;
use Symfony\Component\Process\Process;

/**
 * Video dosyalarını diske kaydeder, ffprobe ile süreyi okur ve
 * ffmpeg ile videodan otomatik thumbnail (1. saniyedeki kare) çıkarır.
 * .NET tarafındaki Kela.Api.Media.MediaProcessor ile aynı davranışı sürdürür.
 */
class MediaProcessor
{
    public const int MAX_VIDEO_MB = 512;

    private const array VIDEO_EXTENSIONS = [
        'mp4', 'webm', 'ogg', 'mov', 'm4v', 'mkv', 'avi', 'mpg', 'mpeg',
    ];

    /** Depolama köküne göre disk dizinleri (local disk = storage/app). */
    public const string VIDEOS_DIR = 'uploads/videos';
    public const string THUMBNAILS_DIR = 'uploads/thumbnails';

    /**
     * Zaten diske kaydedilmiş bir videodan süre ve thumbnail üretir.
     *
     * @return array{thumbnail_path: string|null, duration_seconds: int}
     */
    public function processVideo(string $storedVideoPath): array
    {
        $full = $this->resolveFullPath($storedVideoPath);
        if ($full === null) {
            return ['thumbnail_path' => null, 'duration_seconds' => 0];
        }

        $duration = $this->probeDuration($full);
        $thumbnail = $this->extractThumbnail($full);

        return [
            'thumbnail_path' => $thumbnail,
            'duration_seconds' => $duration,
        ];
    }

    /**
     * DB'deki relative yol → diskteki fiziksel yol.
     *
     * Upload'lar `->store(..., 'local')` ile local diskə yazılır
     * (kök = storage/app/private). Köhnə yazımlar storage/app-in
     * özündə olduğundan, ardıcıllıq üçün əvvəl local diski yoxlayır,
     * tapılmazsa köhnə yola (legacy) da baxır.
     */
    public function resolveFullPath(string $storedPath): ?string
    {
        $path = str_starts_with($storedPath, '/') ? ltrim($storedPath, '/') : $storedPath;

        $candidates = [
            Storage::disk('local')->path($path),   // storage/app/private/uploads/videos/...
            storage_path('app/' . $path),          // legacy: storage/app/uploads/videos/...
        ];

        foreach ($candidates as $candidate) {
            if (is_file($candidate)) {
                return $candidate;
            }
        }

        return null;
    }

    /** ffprobe ile süre (saniye, yuvarlanır). */
    public function probeDuration(string $videoFile): int
    {
        $process = new Process([
            'ffprobe', '-v', 'error',
            '-show_entries', 'format=duration',
            '-of', 'csv=p=0',
            $videoFile,
        ]);
        $process->setTimeout(30);
        $process->run();

        $output = trim($process->getOutput());
        if ($output === '' || ! is_numeric($output)) {
            return 0;
        }

        return (int) round((float) $output);
    }

    /** ffmpeg ile videonun 1. saniyesindeki kareyi thumbnail olarak kaydeder. */
    public function extractThumbnail(string $videoFile): ?string
    {
        $dir = Storage::disk('local')->path(self::THUMBNAILS_DIR);
        if (! is_dir($dir) && ! mkdir($dir, 0777, true) && ! is_dir($dir)) {
            return null;
        }

        $name = Str::uuid()->toString() . '.jpg';
        $thumbFile = $dir . '/' . $name;

        $process = new Process([
            'ffmpeg', '-y',
            '-ss', '1',
            '-i', $videoFile,
            '-frames:v', '1',
            '-q:v', '2',
            $thumbFile,
        ]);
        $process->setTimeout(60);
        $process->run();

        if (! $process->isSuccessful() || ! is_file($thumbFile)) {
            return null;
        }

        return self::THUMBNAILS_DIR . '/' . $name;
    }

    public static function contentTypeFor(string $path): string
    {
        return match (strtolower(pathinfo($path, PATHINFO_EXTENSION))) {
            'mp4' => 'video/mp4',
            'webm' => 'video/webm',
            'ogg', 'ogv' => 'video/ogg',
            'mov' => 'video/quicktime',
            'm4v' => 'video/x-m4v',
            'mkv' => 'video/x-matroska',
            'avi' => 'video/x-msvideo',
            'jpg', 'jpeg' => 'image/jpeg',
            default => 'application/octet-stream',
        };
    }
}
